doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,255 @@
+    - If most changes are not record appearances or disappearances, but existing records changing in a way that they move only a little bit in the ordering. In this case, the output changes only for as many records, that got “jumped over” by the changing record. (Prefix Sum will handle this fine.)
+    - TopK is an important special case: This is when there is a `WHERE ROW_NUMBER() < k`
+        - Instead of doing Prefix Sum, we will transform this into our [efficient TopK implementation](https://www.notion.so/e62fe2b3d8354052ac7d0fe92be1e711).
+
+## Non-goals / Limitations
+
+- We don’t handle such OVER clauses where the ORDER BY inside the OVER is on a String or other complex type. See a discussion on supported types below in the "Types" section.
+- In cases that we handle by Prefix Sum, the groups specified by the composite key of the PARTITION BY and the ORDER BY should be small, see the "Duplicate Indexes" section.
+
+# Details
+
+## Current state
+
+The current way of execution is to put entire partitions into scalars, and execute the window function to all elements by a “scalar aggregation”. This happens in the HIR to MIR lowering, i.e., MIR and LIR don’t know about windowing.
+
+## Proposal
+
+We will use [DD’s prefix_sum](https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/algorithms/prefix_sum.rs) (with some generalizations) for most of the window functions. The bulk of this work will be applied in the rendering, but currently the windowing stuff disappears in the HIR-to-MIR lowering. I propose to create a new relation expression enum variant in both MIR and LIR. This would allow us to focus on windowing for this epic (the epic is quite big already), and then we could have a separate epic for unifying windowing, TopK, and the current Reduce into a “super Reduce”. For a discussion on windowing representations, see the [Alternatives section](https://www.notion.so/Efficient-Window-Functions-87682d6de73a4f6088acde21aba41fe6).
+
+### Rendering
+
+We’ll use three approaches to solve the many cases mentioned in the “Goals” section:
+
+1. We’ll use [DD’s prefix_sum](https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/algorithms/prefix_sum.rs) with some tricky sum functions.
+2. As an extension of 1., we’ll use a generalization of DD’s prefix sum to arbitrary intervals (i.e., not just prefixes).
+3. We transform away the windowing in some special cases (e.g., to TopK, or a simple grouped aggregation + self-join)
+
+We’ll use the word ****index**** in the below text to mean the values of the ORDER BY column of the OVER clause, i.e., they are simply the values that determine the ordering. (Note that it’s sparse indexing, i.e., not every number occurs from 1 to n, but there are usually (big) gaps.)
+
+Now I’ll list all window functions, and how we’ll support them with one of the above approaches:
```

Comment:
In reading this, it starts as an enumeration of window functions, but turns in to a bullet salad. If at all possible, it would help to decouple the moments that are meant to be enumerations from the associated thoughts, to make it easier for the reader to assess the exhaustiveness / find the special case they are looking to understand.

Reply:
Yes, sorry, I didn't realize how annoying can a bullet salad be. I completely changed the structuring. The "How to handle each window function" section has numbered subsections, enumerating all window functions (grouped into categories). There are also two big horizontal lines delineating this big enumeration section.